htdocs/core/modules/expedition/doc/pdf_espadon.modules.php

```diff
@@ -991,10 +991,27 @@ protected function _pagehead(&$pdf, $object, $showaddress, $outputlangs)
 			$posx = $this->marge_gauche + 3;
 		}
 		//$pdf->Rect($this->marge_gauche, $this->marge_haute, $this->page_largeur-$this->marge_gauche-$this->marge_droite, 30);
-		if (isModEnabled('barcode')) {
-			// TODO Build code bar with function writeBarCode of barcode module for sending ref $object->ref
-			//$pdf->SetXY($this->marge_gauche+3, $this->marge_haute+3);
-			//$pdf->Image($logo,10, 5, 0, 24);
+		if (isModEnabled('barcode') && getDolGlobalString('BARCODE_RECEPTION_INVOICE')) {
+			require_once DOL_DOCUMENT_ROOT.'/core/modules/barcode/doc/tcpdfbarcode.modules.php';
+
+			$encoding = 'QRCODE';
+			$module = new modTcpdfbarcode($this->db);
+			if ($module->encodingIsSupported($encoding)) {
+				$result = $module->writeBarCode($object->ref, $encoding);
+
+				// get path of qrcode image
+				$newcode = $object->ref;
+				if (!preg_match('/^\w+$/', $newcode) || dol_strlen($newcode) > 32) {
+					$newcode = dol_hash($newcode, 'md5');
+				}
+				$barcode_path = $conf->barcode->dir_temp . '/barcode_' . $newcode . '_' . $encoding . '.png';
+			}
+
+			if ($result > 0) {
+				$pdf->Image($barcode_path, $this->marge_gauche,  $this->marge_haute +80 -5, 20, 20);
```

Comment:
Why not $pdf->write2DBarcode ?

Reply:
> Why not $pdf->write2DBarcode ?

We want to get barcode using an intermediate image using >writeBarCode() because >writeBarCode() is an abstraction class that is common to different barcode generator (phpbarcode and tcpdfbarcode and more in future if we need).

Reply:
Made the remark because seems more complicated to make first png using TCPDF and then use image iso of using TCPDF to write barcode in pdf. For object->ref not big issue, but if it is a big QR, QR quality coud be less by the image resize.

Reply:
> Made the remark because seems more complicated to make first png using TCPDF and then use image iso of using TCPDF to write barcode in pdf. For object->ref not big issue, but if it is a big QR, QR quality coud be less by the image resize.

Not the same if you use $pdf->write2DBarcode() ? I suppose that tcpdf also generated an intermediate image or the barcode is directly saved as pdf instruction ?

Reply:
$pdf->write2DBarcode() draws barcode pixel on pdf: https://github.com/Dolibarr/dolibarr/blob/dc8b79e0d4e3a9ac1eef56c1a5b8bb407d825cae/htdocs/includes/tecnickcom/tcpdf/tcpdf.php#L16071
No intermediate image involved.
$modTcpdfbarcode->writeBarCode() is more for showing barcode on screen.

Reply:
You're right. We should enhance this by adding a method
`function pdf_barcode(&$pdf, ...)`
into file **pdf.lib.php**
with all content inside the 
`if (isModEnabled('barcode') && getDolGlobalString('BARCODE_RECEPTION_INVOICE')) {
...
}`

So we will be able to replace current code with a more opmitised version